### PR TITLE
PEN-1437 & PEN-1440: Navbar - Horizontal Links & Logo Alignment

### PR DESCRIPTION
**Pull request by vinnie-olsen**
## Description
Added functionality for "Horizontal Links" populated through the `Horizontal Links Hierarchy` custom field. Also added the ability to specify the alignment of the nav logo - configurable through the `Logo Alignment` custom field.

## Jira Ticket
- [PEN-1437](https://arcpublishing.atlassian.net/browse/PEN-1437)
- [PEN-1440](https://arcpublishing.atlassian.net/browse/PEN-1440)

## Acceptance Criteria
**PEN-1437 ---**
1. A new Custom Field is added to the Navigation Chain called Logo alignment according to the specs above 
a. The default value should be center, including if no value is set for this Custom Field at all (so that existing placements do not break) 
b. If the value is set to left, then the logo should be left aligned within the navigation, according to the design mockup
c. This custom field should be in a group called Logo
2. Tests and documentation are updated to cover this functionality 

**PEN-1440 ---**
1. A new Custom Field called Horizontal Links hierarchy is added to the Navigation Chain per the specifications above 
2. If a PageBuilder user enters a hierarchy in this custom field, the items within that hierarchy should render as links in the navigation bar, on the Desktop breakpoint 
a. Only the top level of items in this hierarchy should render (if there are children, they are not expected to render)
b. Link creation should follow the practices we’ve established elsewhere for Site Service hierarchies, including: 
i. Both Link objects and Section objects should render 
ii. If the URL is relative, it should open on the same tab; if the URL is absolute, it should open in a new tab 
c. If list of links is longer than space allows, some of the links should be hidden from the reader’s view behind the components on the right side of the navigation
d. The horizontal links do not render at all on the tablet or mobile breakpoints
3. If a PageBuilder user enters a hierarchy in the Horizontal Links hierarchy but the Logo alignment is set to center (from PEN-1437: [CMG] Ability to left-align logo in navigation chainIN PROGRESS, then the horizontal links should not appear 
a. A message can appear in the PageBuilder admin (only – not on the published page) that tells the user: “In order to render horizontal links, the logo must be aligned to the left”) 
4. Tests and documentation are updated to cover this functionality

## Test Steps
1. In `Fusion-News-Theme` repo, checkout `master` & `git pull`
2. After checking out this feature branch in `fusion-news-theme-blocks`, run `rm -rf ./node_modules && npx lerna clean && npm i && npx lerna bootstrap && cd blocks/header-nav-chain-block && npm i && cd ../..`
3. Run `npx fusion start -f -l @wpmedia/header-nav-chain-block,@wpmedia/shared-styles` in `Fusion-News-Theme`
4. Open 'Homepage' in PB editor, edit or add `Navigation - Arc Chain` feature to the page
5. Open this feature in PB editor, add a `Horizontal Links Hierarchy` custom field of `default`
6. Still in PB editor for the nav feature, change the `Logo Alignment` custom field to `left`
7. You should see a logo displaying on the left side of the nav bar (to the right of the search and menu buttons). You should also see the horizontal links rendering in the middle section of the nav bar.
8. Stage, publish & open http://localhost/pf/homepage/?_website=dagen in a new tab. Check that the nav bar displays as expected (nav logo left, horizontal links in middle). Check that this displays properly across all breakpoints.
9. Go back to PB editor, click on nav feature to open in editor, change `Logo Alignment` custom field to `center`. Then stage & publish again.
10. You should see that the logo now displays in the middle section of the nav bar and the horizontal links get hidden. In PB admin, you'll see a warning banner letting you know that horizontal links cannot be displayed while logo is centered. This will only appear of logo has `center` alignment and horizontal links custom field is populated with a hierarchy.
11. Check that this displays properly across breakpoints.
12. Feel free to test any other sites with the above steps for additional verification (different logos, different site hierarchy taxonomy). 

### Before
There was no ability to display a horizontal link box in the nav bar and the logo was automatically center-aligned.

### After
You can now display a list of horizontal links (populated from site hierarchy) and configure the logo to be either left or center-aligned.

## Dependencies or Side Effects
_Examples of dependencies or side effects are:_
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Update to css framework or SDK
- Dependency on another PR that needs to be merged first

## Review Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [X] Confirmed all the test steps above are working
- [X] Confirmed there are no linter errors
- [X] Confirmed this PR has reasonable code coverage
  - [X] Confirmed this PR has unit test files
  - [X] Ran `npm test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [X] Confirmed relevant documentation has been updated/added.

### CODE REVIEW

**Review by rachelpost — APPROVED** [review]

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+
+function getLocation(uri) {
+  let url;
+  if (typeof window === 'undefined') {
+    url = new URL(uri, 'http://example.com');
```
Comment (JackHowa) [inline blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:8]: not sure I totally understand `example.com` included here as the base? for testing. for ssr?

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  } else {
+    url = document.createElement('a');
+    // IE doesn't populate all link properties when setting .href with a relative URL,
```
Comment (JackHowa) [inline blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:11]: fun fun IE -- good thinking

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx [hunk 3]
```diff
@@ -220,6 +245,17 @@ Nav.propTypes = {
...
     }),
     signInOrder: PropTypes.number,
+    logoAlignment: PropTypes.oneOf([
+      'center', 'left',
+    ]).tag({
+      label: 'Logo alignment',
+      group: 'Logo',
+      defaultValue: 'center',
+    }),
+    horizontalLinksHierarchy: PropTypes.string.tag({
```
Comment (JackHowa) [inline blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx:255]: are you handling the default value for this for preexisting pages?

File: blocks/shared-styles/scss/_header-nav.scss [hunk 4]
```diff
@@ -145,8 +145,10 @@ body.nav-open {
   align-items: center;
   display: flex;
   justify-content: space-between;
+  // justify-content: flex-start;
```
Comment (JackHowa) [inline blocks/shared-styles/scss/_header-nav.scss]:
```suggestion
```

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx (hunk 1 above)
Comment (rachelpost) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:8]: yes - this was included in the copyover from the links bar block and is used for testing.

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx (hunk 3 above)
Comment (rachelpost) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx:255]: For preexisting pages the horizontal links won't be shown - this custom field is used to populate the links but also acts as the on/off switch to display the horizontal links.  so no horizontalLinksHierarchy => no horizontal links shown.  The default is null hierarchy and thus no horizontal links shown which is what the preexisting components will do.

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx (hunk 1 above)
Comment (JackHowa) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:8]: can you remove it if it's just used for testing? sorry, it just seems unnecessary to generate this for all ssr renders

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx (hunk 3 above)
Comment (JackHowa) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/default.jsx:255]: oh ok, we're relying on it being null and evaluating as falsey. If it was up to me, I'd always want a default value just so  we can be explicit about this (it's bitten me before).

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/default.test.jsx [hunk 5]
```diff
@@ -105,6 +105,43 @@ describe('the header navigation feature for the default output type', () => {
...
+        expect(navLogoEl.hasClass('nav-logo-center')).toBe(true);
+      });
+    });
+
+    describe('when the nav logo is left-aligned', () => {
+      it('should be rendered inside "nav-left" container (left-aligned)', () => {
+        getProperties.mockImplementation(() => ({ primaryLogo: 'https://test.com/my-nav-logo.svg' }));
+        const wrapper = mount(<Navigation customFields={{ logoAlignment: 'left' }} />);
+        expect(wrapper.props().customFields.logoAlignment).toBeDefined();
+        expect(wrapper.props().customFields.logoAlignment).toEqual('left');
```
Comment (JackHowa) [inline blocks/header-nav-chain-block/chains/header-nav-chain-block/default.test.jsx:138]: good tests

File: blocks/shared-styles/scss/_header-nav.scss [hunk 6]
```diff
@@ -145,8 +145,10 @@ body.nav-open {
   align-items: center;
   display: flex;
   justify-content: space-between;
```
Comment (JackHowa) [inline blocks/shared-styles/scss/_header-nav.scss:147]: is there a reason why this was added as commented out?

**Comment by JackHowa** [comment]
@rachelpost @vinnie-olsen any screenshots or visual ways of showing this work?

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx (hunk 1 above)
Comment (kascote) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:8]: the URL object need a base domain for work. the URL is used to parse the url on `fixTrailingSlash` function to do some checks. The final url is never used.

**Comment by JackHowa** [comment]
fantastic test instructions again @vinnie-olsen

**Comment by JackHowa** [comment]
Design question: should the positioning 'Horizontal hierarchy' custom field be a dropdown not a custom input field ?

Screen Shot 2020-11-13 at 09 57 07

**Review by JackHowa — APPROVED** [review]

File: blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx (hunk 1 above)
Comment (JackHowa) [reply blocks/header-nav-chain-block/chains/header-nav-chain-block/_children/horizontal-links/_children/link.jsx:8]: ok if this is absolutely necessary

**Comment by vinnie-olsen** [comment]
@JackHowa for the `Horizontal Links hierarchy` field, it looks like it is supposed to be just a text field and not a dropdown per the requirements in https://arcpublishing.atlassian.net/browse/PEN-1440

**Comment by vinnie-olsen** [comment]
@JackHowa thanks for reviewing this for us! 🙂

**Comment by JackHowa** [comment]
> @JackHowa for the Horizontal Links hierarchy field, it looks like it is supposed to be just a text field and not a dropdown per the requirements in https://arcpublishing.atlassian.net/browse/PEN-1440

cool, just checking

**Pull request merged**